pkgs/build-support/compress-drv/default.nix

```diff
@@ -0,0 +1,71 @@
+/*
+compressDrv compresses files in a given derivation.
+
+Inputs:
+
+- formats :: [String]
+
+    List of file extensions to compress.
+
+    Example: ["txt" "svg" "xml"]
+
+- compressors :: [String]
+
+    A list of compressor names to use. Each element will need to have
+    an associated compressor in the same arguments (see below).
+
+    Example: ["gz" "zstd"]
+
+- compressor-<EXTENSION> :: String
+
+    Map a desired extension (e.g. `gz`) to a compress program.
+
+    The compressor program that will be executed to get the `COMPRESSOR`
+    extension. The program should have a single " {}", which will be the
+    replaced with the target filename.
+
+    Compressor must:
+    - read symlinks (thus --force is needed to gzip, zstd, xz).
+    - keep the original file in place (--keep).
+
+    Example compressor:
+
+      compressor-xz = "${xz}/bin/xz --force --keep {}";
+
+ See compressDrvWeb, which is a wrapper on top of compressDrv, for broader
+ use examples.
+*/
+{
+  lib,
+  runCommand,
+}: drv: {
+  formats,
+  compressors,
+  ...
+} @ args: let
+  validProg = ext: prog: let
+    matches = (builtins.length (builtins.split "\\{}" prog) - 1) / 2;
+  in
+    lib.assertMsg
+    (matches == 1)
+    "compressor-${ext} needs to have exactly one '{}', found ${builtins.toString matches}";
+  compressorMap = lib.filterAttrs (k: _: (lib.hasPrefix "compressor-" k)) args;
+  mkCmd = ext: prog:
+    assert validProg ext prog; ''
+      find -L $out -type f -regextype posix-extended -iregex '.*\.(${formatsPipe})' -print0 \
+        | xargs -0 -P$NIX_BUILD_CORES -I{} ${prog}
+    '';
+  formatsPipe = builtins.concatStringsSep "|" formats;
+in
+  runCommand "${drv.name}-compressed" {} ''
+    mkdir $out
+    (cd ${drv}; find -L -type d -exec mkdir -p $out/{} ';')
+    (cd ${drv}; find -L -type f -exec ln -s ${drv}/{} $out/{} ';')
```

Comment:
I love to see `lndir` gone here.

Reply:
It's not bad per se -- it's just part of the the `xorg` world, and replacing it with something that's natively in the `stdenvNoCC` is both simpler and requires less learning.

Reply:
I actually like `lndir` more, because it's quite widely used, a single operation (instead of two traversals) and automatically resolves nested symlinks (which the previous PR version didn't). I just couldn't get it to work before, and no longer understand/remember why.

Reply:
I am not aware of any tool replacing lndir